### PR TITLE
Add x86_64-musl as a host architecture

### PR DESCRIPTION
**Pull request by strfry**
rustc is already shipped on Alpine Linux, this should be an easy thing to also enable it for the rustup channels.

This PR contains an additional commit for testing the build with Travis, i will remove it once that worked.

### CODE REVIEW

**Comment by alexcrichton** [comment]
Thanks for the PR! One of the more difficult problems to deal with here is that we'll need a C++ compiler that can target musl. Currently the container only has a C complier, but the build will otherwise fail due to the lack of a C++ compiler which means we won't be able to build LLVM against musl instead of glibc

**Comment by strfry** [comment]
The current build failure is apparently related to #44991

@alexcrichton I think the container is equipped with a native c++ compiler, and musl-gcc will happily use that when instructed so with --std . As you can see, it already goes way into stage1

**Comment by alexcrichton** [comment]
@strfry yes that's an other issue where you'll want to configure musl to be compiled dynamically rather than statically. I don't believe that the c++ compiler in the container works, though, you'll need a C++ standard library compiled against the musl C standard library

**Comment by strfry** [comment]
@alexcrichton I temporarily applied the fix suggested in #44991 via RUSTFLAGS (that should be fixed with another PR, like suggested in that issue).
Now it got to the point where the C++ stdlib for musl is missing.
I suppose that's the end for the musl-gcc hackery, which must be replaced by a proper musl-toolchain.sh script. This might also enable building for other architectures than x86_64 relatively easy as a corollary.

**Comment by corbinu** [comment]
@strfry Anything I can do to help here? I am super excited about this

**Comment by TimNN** [comment]
Ping from triage @strfry: What is the status of this PR?

**Comment by strfry** [comment]
@TimNN The development has been stalled as my build machine that i use as a Docker host to reproduce the Travis Build locally died. Now i repaired it and hopefully can continue fixing those late-stage build issues.

@corbinu I'm a bit out of the loop of what is the current actual problem. There a lot of new info in #55566 . 
One thing i remember postponing is the `ENV RUSTFLAGS="-C target-feature=-crt-static"` hack, that eventually needs to be replaced with a proper fix, to make musl rustc  by default build dylib's as dynamic objects.

**Comment by strfry** [comment]
@Mark-Simulacrum Apparently, musl-rustc isn't accepting crate of type "proc-macro". This might be related to the root cause of why it's refusing to create dylib's without `-crt-static`.
I'm still struggling with my build setup to be able to trace into the related code.

**Comment by smaeul** [comment]
@strfry There's a `crt-static` option you should set in `config.toml` (see the example file), that will set the `crt-static` feature at the appropriate places in the bootstrap. You shouldn't need to touch `RUSTFLAGS` manually.

For the testsuite, you'll need something like https://github.com/smaeul/rust/commit/294a9feda6b101466e9ca1646bc3c802afcbe36a for `proc-macro` crates to build.

File: src/ci/docker/scripts/musl-toolchain.sh [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+OUTPUT=/usr/local
+shift
+
+git clone https://github.com/richfelker/musl-cross-make -b v0.9.7
+cd musl-cross-make
+
+hide_output make -j$(nproc) TARGET=$TARGET
+hide_output make install TARGET=$TARGET OUTPUT=$OUTPUT
+
+cd ..
```
Comment (sanmai-NL) [inline src/ci/docker/scripts/musl-toolchain.sh]: `cd -` would be more general.

File: src/ci/docker/scripts/musl-toolchain.sh [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+          -DLLVM_PATH=/build/llvm-release_$LLVM \
+          -DLIBUNWIND_ENABLE_SHARED=0 \
+          -DCMAKE_C_COMPILER=$CC \
+          -DCMAKE_CXX_COMPILER=$CXX \
+          -DCMAKE_C_FLAGS="$CFLAGS" \
+          -DCMAKE_CXX_FLAGS="$CXXFLAGS"
+
+hide_output make -j$(nproc)
+cp lib/libunwind.a $OUTPUT/$TARGET/lib
+cd ../ && rm -rf libunwind-build
```
Comment (sanmai-NL) [inline src/ci/docker/scripts/musl-toolchain.sh]: Idem.

**Comment by TimNN** [comment]
Ping from triage @strfry: It looks like your PR fails on travis.

**Comment by mati865** [comment]
@strfry I believe I know what is wrong.
I'm not sure how to test it, could you amend this commit https://github.com/mati865/rust/commit/53eb5d9b1c0c1e9c872a149c7be46973a3398d12?

**Comment by mati865** [comment]
My bad, force-pushed fixed commit: https://github.com/mati865/rust/commit/f9b551be2c322bcef479867d75ac03dcd6856111
You can fix it in any way you like.

File: src/librustc_target/spec/linux_musl_base.rs [hunk 3]
```diff
@@ -46,7 +46,7 @@ pub fn opts() -> TargetOptions {
     base.has_elf_tls = true;
     base.has_rpath = true;
     base.position_independent_executables = true;
-    base.relro_level = true;
+    base.relro_level = RelroLevel::Full;
```
Comment (mati865) [inline src/librustc_target/spec/linux_musl_base.rs]: Also change import to include `spec::{LinkerFlavor, TargetOptions, RelroLevel}`;

**Comment by mati865** [comment]
One of the blockers is https://github.com/rust-lang/rust/issues/53968

**Comment by alexcrichton** [comment]
I'm gonna close this as it's gone a bit stale now and I think it still needs some work, but it can of course be resubmitted!

**Comment by farcaller** [comment]
@alexcrichton if this is closed can we have a tracking bug for musl instead please, so that the interested parties could subscribe to it?

**Comment by alexcrichton** [comment]
Of course! You can feel free to open a bug

**Pull request closed without merging**